### PR TITLE
Proposal for #313

### PR DESCRIPTION
**Pull request by Randelung**
It's now 6am, so I'm just dropping this and heading to bed.
Feel free to comment, the thing's been running for an hour now without issues. I think the concept holds.
So apparently the #313 doesn't work in titles.

### CODE REVIEW

File: asyncua/common/connection.py [hunk 1]
```diff
@@ -303,7 +303,7 @@ def _check_incoming_chunk(self, chunk):
                 else:
                     # Condition for monotonically increase is not met
                     raise ua.UaError(f"Received chunk: {chunk} with wrong sequence expecting:"
-                                     f" {self._peer_sequence_number}, received: {seq_num},"
+                                     f" {self._peer_sequence_number + 1}, received: {seq_num},"
```
Comment (Randelung) [inline asyncua/common/connection.py:306]: It's supposed to be one more than the currently tracked number, which is also the check just above. I got a little confused when the error popped up and said "expecting: 594, received: 594" 😀

**Comment by swamper123** [comment]
Tracvis CI seems to use an pypy3 version which seems to act like Python 3.6.9. ... *sigh*
So some stuff may not work here.

The `send_hello` stuff is breaking, caused by a timeout. Could this be caused by the Lock?

**Comment by Randelung** [comment]
Oh wow, ok, not sure. I've never worked with pypy before. The locks by themselves shouldn't be an issue since they've been [stable for a while](https://docs.python.org/3.5/library/asyncio-sync.html#asyncio.Lock). I'll try to get a pypy environment going I guess.
Do I interpret correctly that the other three tests with normal python tested fine?

File: asyncua/client/ua_client.py [hunk 2]
```diff
@@ -50,42 +51,50 @@ def connection_lost(self, exc):
...
+                    header = header_from_binary(buffer_header)
                 except ua.utils.NotEnoughData:
                     self.logger.debug('Not enough data while parsing header from server, waiting for more')
-                    self.receive_buffer = data
                     return
-                if len(buf) < header.body_size:
-                    self.logger.debug(
-                        'We did not receive enough data from server. Need %s got %s', header.body_size, len(buf)
+                if len(buffer_header) < header.body_size:
+                    self.logger.warning(
```
Comment (Randelung) [inline asyncua/client/ua_client.py:73]: Whoops, this is supposed to be debug. I had it set higher for better visibility.

Comment (Randelung) [reply asyncua/client/ua_client.py:73]:
```suggestion
                    self.logger.debug(
```

**Comment by Randelung** [comment]
Are you sure it's 3.6.9? I got pypy3.6-v7.3.2 set up, but it says no module named dataclasses for crypto/uacrypto.py, so I assume this wouldn't have worked for a while.
Using pypy3.7-v7.3.2 I can't get my application to run, it hangs on initializing the server (whose source I didn't touch). CTRL-C won't even work to end it. The master branch has the same behavior. It might just be pypy3...
The prosys example crashes on master and runs on fix_313 (except it's noticeably slower than regular python).

**Comment by Randelung** [comment]
Another consideration to be made is whether the asyncio.Protocol approach is best or if a `reader, writer = await asyncio.open_connection(...)` would yield better results. The loop over the `reader` would dictate when packets are being read and processed, so parsing can be offloaded into a different task only if the channel state allows it. (Since dataclasses are being used I'm just also going with the minimum of 3.7.)

**Comment by swamper123** [comment]
> Are you sure it's 3.6.9?

From their page they say:

> PyPy3 implements the Python language version 3.6.9. It has been released, but Python is a large language and it is quite possible that a few things are missing.

grafik

So errors like this may happen:

grafik
(the last line is caused by "Python 3.6") Well this is my observation on this point of the Test.

It seems that the la[test Versions for ppypy3.7 still got some problems ](https://buildbot.pypy.org/summary?branch=%3Ctrunk%3E) among others with asyncio. So we should be patient about that perhaps.

**Comment by Randelung** [comment]
How do you propose we continue? Anything you need me to do?

**Comment by swamper123** [comment]
I have run it with Python 3.6 and it worked.
I tried to get pypy3 running (and I failed with that 😞), so I couldn't test your branch localy.

> How do you propose we continue? Anything you need me to do?

To be honest, I don't know. The last commit of the master branch ran fine with pypy3. 😕 Travis is even running 1:1 the same version as your testbuild did. I kicked the test again and we may see, if something strange happens, but I believe not.

**Comment by Randelung** [comment]
> I have run it with Python 3.6 and it worked.
Wait what? How? Didn't dataclasses arrive in 3.7? server.users and crypto.uacrypto both use dataclasses. asyncio.create_task is 3.7+. Am I missing something?

Pypy was a pain to get running tbh. Took ages to install and compile dependencies and dependency dependencies (am on Windows, hardly any of the packages just installed as promised on the pypy page). I couldn't get the master branch to run the example on 3.6, that could be a Windows problem, though.

**Comment by swamper123** [comment]
Nothing changed... I am out of ideas here. Sorry for that 😕

**Comment by swamper123** [comment]
> Wait what? How? Didn't dataclasses arrive in 3.7? server.users and crypto.uacrypto both use dataclasses. asyncio.create_task is 3.7+. Am I missing something?

Clone your repo, checkout branch fix_313, execute `python3.6 pytest -v -s` in cmd...so localy.

> Didn't dataclasses arrive in 3.7?

Yes, but there is a backport on pypy and dataclasses is (still, but perhaps shouldn't anymore) in `.travis` install.

> Pypy was a pain to get running tbh.

Python and Windows are enemies since I can thought...and I bet it is even worse using pypy ... and more worse run pytest in pypy...and even more worse test asyncio stuff with pytest in pypy on windows (asyncio (python around 3.6) and windows haven't been good friends, don't know if this is solved atm).

**Comment by Randelung** [comment]
> Yes, but there is a backport on pypi for python3.6 and dataclasses

Oooooh, that explains it. I'll replace the create_task with the loop version, then, and try again!
And then I'll try the suggestion that oroulet made about moving the open_secure_channel code into the receive loop.

File: asyncua/client/ua_client.py [hunk 3]
```diff
@@ -50,42 +51,50 @@ def connection_lost(self, exc):
         self.transport = None
 
     def data_received(self, data: bytes):
-        if self.receive_buffer:
-            data = self.receive_buffer + data
-            self.receive_buffer = None
-        self._process_received_data(data)
+        # buffer everything
+        self.receive_buffer += data
+        asyncio.create_task(self._async_data_received())
```
Comment (Randelung) [inline asyncua/client/ua_client.py:56]: Not 3.6 compatible

Comment (Randelung) [reply asyncua/client/ua_client.py:56]:
```suggestion
        self.loop.create_task(self._async_data_received())
```

**Comment by Randelung** [comment]
Um I assume it's displaying the previous result. I'm getting 1 failed, 461 passed, 5 skipped, 6 warnings, 1 error for both pypy versions now.

**Comment by swamper123** [comment]
> Um I assume it's displaying the previous result. I'm getting 1 failed, 461 passed, 5 skipped, 6 warnings, 1 error for both pypy versions now.

The test was cancelled. I kicked it again

**Comment by Randelung** [comment]
Now two tests are running :D The cancelled one was the previous commit I think.

**Comment by oroulet** [comment]
another fix has been merged

**Pull request closed without merging**